File: code/ch18/18.3.1.java
class UnionFind {
    // 每个节点的父节点
    int[] parent;
    // 以该节点为根的树权值（树高度）
    int[] rank;
    // 连通区域数量
    int cnt;
    
    public UnionFind(int n) {
        this.parent = new int[n];
        this.rank = new int[n];
        this.cnt = n;
        for (int i = 0; i < n; i++) {
            this.parent[i] = i;
            this.rank[i] = 0;
        }
    }

    public int find(int p) {
        while (p != this.parent[p]) {
            p = this.parent[p];
        }
        return p;
    }

    public void union(int p, int q) {
        int root_p = this.find(p), root_q = this.find(q);
        if (root_p == root_q) {
            return;
        }
        if (this.rank[root_p] > this.rank[root_q]) {
            this.parent[root_q] = root_p;
        } else if (this.rank[root_p] < this.rank[root_q]) {
            this.parent[root_p] = root_q;
        } else {
            this.parent[root_p] = root_q;
            this.rank[root_q] += 1;
        }
        this.cnt -= 1;
    }
}

File: code/ch18/18.3.1.py
class UnionFind:
    def __init__(self, n: int):
        # 每个节点的父节点
        self.parent = [i for i in range(n)]
        # 以该节点为根的树权值（树高度）
        self.rank = [0 for i in range(n)]
        # 连通区域数量
        self.cnt = n

    def find(self, p: int) -> int:
        while p != self.parent[p]:
            p = self.parent[p]
        return p

    def union(self, p: int, q: int):
        root_p, root_q = self.find(p), self.find(q)
        if root_p == root_q:
            return
        if self.rank[root_p] > self.rank[root_q]:
            self.parent[root_q] = root_p
        elif self.rank[root_p] < self.rank[root_q]:
            self.parent[root_p] = root_q
        else:
            self.parent[root_q] = root_p
            self.rank[root_p] += 1
        self.cnt -= 1
